Add always trimmed attributes list for browser config

marek-safar opened the pull request:
Example of size reduction on SPC with the most minimalistic Hello
World console sample.

| System.Private.CoreLib | Size (kB) | Delta |
|-|-|-|
| Trimmed Before | 1358 kB |  |
| Trimmed After | 1271 kB | **-87 kB** |


Trimming the [CarChecker app](https://github.com/SteveSandersonMS/CarChecker/tree/net5) with this change results in 

| CarChecker | Size | Delta |
|-|-|-|
| Trimmed Before | 3,584,512 bytes|  |
| Trimmed After | 3,474,432 bytes | **-108 kB** |
| Trimmed After with removing CompilerGenerated | 3,453,952 bytes | **-128 kB** |

@eerhardt @vitek-karas

marek-safar committed:
Add always trimmed attributes list for browser config

Example of size reduction on SPC with the most minimalistic Hello
World sample.

| System.Private.CoreLib | Size (kB) | Delta |
|-|-|-|
| Original | 1358 kB | - |
| Trimmed | 1271 kB | - 87 kB |

eerhardt commented on src/mono/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -152,6 +152,9 @@
     <ILLinkSubstitutionsXmls Include="$(ILLinkDirectory)ILLink.Substitutions.$(Platform).xml"
                              Condition="Exists('$(ILLinkDirectory)ILLink.Substitutions.$(Platform).xml')" />
+
+    <ILLinkLinkAttributes Include="$(ILLinkDirectory)/ILLink.LinkAttributes.$(Platform).xml"
```
Nothing is using this MSBuild Item, so it isn't doing anything yet. We need to add some logic to `eng\illink.targets` to use it.

jkotas commented on src/mono/netcore/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.wasm.xml:
```diff
@@ -0,0 +1,142 @@
+    <type fullname="System.ObsoleteAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.ParamArrayAttribute">
```
`ParamArrayAttribute` is considered by reflection for method binding. This will affect how reflection Invoke behaves.

marek-safar replied:
Didn't think about that, removed

jkotas commented on src/mono/netcore/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.wasm.xml:
```diff
@@ -0,0 +1,142 @@
+    <type fullname="System.Runtime.CompilerServices.CompilerGeneratedAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.Runtime.CompilerServices.DynamicAttribute">
```
`DynamicAttribute` is considered by Microsoft.CSharp. Stripping it will affect how `dynamic` keyword behaves.

marek-safar replied:
Good catch!

stephentoub commented on src/mono/netcore/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.wasm.xml:
```diff
@@ -0,0 +1,142 @@
+    </type>
+    <type fullname="System.Diagnostics.CodeAnalysis.MemberNotNullWhenAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
```
Duplicate?

marek-safar replied:
Removed

marek-safar committed:
Review feedback

stephentoub commented on src/mono/netcore/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.wasm.xml:
```diff
@@ -0,0 +1,142 @@
+    </type>
+    <type fullname="System.Runtime.CompilerServices.CompilerGeneratedAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
```
CompilerGenerated is checked by StackTrace.ToString, so this will affect Exception.StackTrace/ToString(), for example.

marek-safar replied:
Hmm, that's ugly. Looking at the code it looks like CompilerGeneratedAttribute usage there is optimization like check. 

We could remove the check or if it's really needed we could remove this attribute under Debug.IsEnabled condition only.

eerhardt replied:
Should we remove this one and add it with the rest of the Debug-only attributes under the `Debugger.IsSupported` feature switch?

Examples of the Debug-only attributes that can be removed are:

* DebuggerTypeProxyAttribute
* DebuggerDisplayAttribute

stephentoub replied:
> Looking at the code it looks like CompilerGeneratedAttribute usage there is optimization like check.

I think it's more there in case the types in question were implemented by someone manually; it's trying to only remove the relevant frames when they're on compiler generated types for async and iterators.  It may be over protective, but I'm hesitant to remove it for the purposes of this PR.

> Should we remove this one and add it with the rest of the Debug-only attributes under the Debugger.IsSupported feature switch?

That sounds reasonable.

marek-safar replied:
Changed it to be conditional

akoeplinger replied:
just to make sure I understand: in cases where debugger is disabled the StackTrace would now include the compiler-generated frames?

stephentoub replied:
Correct

stephentoub commented on src/mono/netcore/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.wasm.xml:
```diff
@@ -0,0 +1,133 @@
+    </type>
+    <type fullname="System.Runtime.CompilerServices.IntrinsicAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
```
Is Intrinsic not used by the runtime?

marek-safar replied:
It's not used by MonoVM/Interpreter

vitek-karas commented on src/mono/netcore/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.wasm.xml:
```diff
@@ -0,0 +1,133 @@
+    <type fullname="System.CLSCompliantAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.ObsoleteAttribute">
```
I suggest we add comment about things which may get broken by this - to document the fact that we understood it and are willing to live with that change of behavior.
In this case we know that XmlSerializer is using this attribute in some cases.

eerhardt replied:
Done.

vitek-karas commented on src/mono/netcore/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.wasm.xml:
```diff
@@ -0,0 +1,133 @@
+<linker>
+  <assembly fullname="*">
```
Can we limit the assembly "*" to only the cases where it's necessary? it's relatively expensive for the linker (as it has to search all assemblies in the closure and find all occurrences of the type).

For most of these attributes we know the assembly they come from, so we should be specific.

I personally would also vote for adding the XML to the same assembly the attribute is from (just so that we don't have unnecessary cross assembly references, especially the "wrong" way from CoreLib upwards).

marek-safar replied:
yeah, we could split the types which are not compiler-generated into its own section

eerhardt replied:
Done.

eerhardt committed:
Combine and embed the ILLink.LinkAttributes.xml file into the assembly.

eerhardt commented on eng/illink.targets:
```diff
@@ -186,6 +209,8 @@
       <!-- don't remove the embedded substitutions xml resource since ILLink may run again on the assembly -->
       <ILLinkArgs Condition="'$(ILLinkSubstitutionsXml)' != ''">$(ILLinkArgs) --strip-substitutions false</ILLinkArgs>
+      <!-- don't remove the embedded link attributes xml resource since ILLink may run again on the assembly -->
+      <ILLinkArgs Condition="'$(ILLinkLinkAttributesXml)' != ''">$(ILLinkArgs) --strip-link-attributes false</ILLinkArgs>
```
@marek-safar @vitek-karas @tlakollo - When we run the linker during the `dotnet/runtime` build, how does the linker know to ignore this LinkAttributes.xml file? We don't want to strip the attributes out during the `dotnet/runtime` build, but only when the app is being linked.

vitek-karas replied:
`--strip-link-attributes false` & `--ignore-link-attributes true`

eerhardt committed:
Update linker command line to ignore attributes file.
Add ExcludeFromCodeCoverageAttribute.

eerhardt committed:
PR feedback.
Add note about Obsolete.
Only use assembly fullname="*" for compiler generated attributes.

marek-safar committed:
Update ILLink.LinkAttributes.wasm.xml

jkotas commented on src/mono/netcore/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.wasm.xml:
```diff
@@ -0,0 +1,149 @@
+    </type>
+    <type fullname="System.Runtime.CompilerServices.CompilerGeneratedAttribute" feature="System.Diagnostics.Debugger.IsSupported" featurevalue="false">
+      <!--
+        The attribute is used during pretty stack trace printing
```
`StackTraceHidden` is used for the same purpose. You can add it with the same condition.

marek-safar replied:
Yep, there is a bunch of others which should be added as conditional (e.g. whole ETW). I'd prefer to add them in follow up PRs

eerhardt commented:
Setting `No Merge` on this PR because as soon as we do, it will break publishing any Blazor application until https://github.com/mono/linker/issues/1341 is fixed.

eerhardt commented:
Removing the `No Merge` label now that the fix for https://github.com/mono/linker/issues/1341 has been merged into dotnet/sdk.

eerhardt reviewed (approved):
I think this is ready to be merged. Any other feedback?
